sdk/tests/conformance2/extensions/ext-texture-norm16.html

```diff
@@ -49,49 +105,77 @@
   }
 }
 
-function testNorm16Texture(internalFormat, format, type) {
-  let pixelValue;
-  let expectedValue;
-  let imageData;
+function texImageSrc(internalFormat, format, type) {
+  debug("texFunc: texImageSrc");
+  gl.texImage2D(gl.TEXTURE_2D, 0, internalFormat, 1, 1, 0, format, type, imageSrc[type]);
+}
+
+function texImageBitmap(internalFormat, format, type) {
+  if(!window.createImageBitmap || !window.ImageBitmap) {
+    debug("ImageBitmap isn't supported");
+    return;
+  }
+  debug("texFunc: texImageBitmap");
+  gl.texImage2D(gl.TEXTURE_2D, 0, internalFormat, 1, 1, 0, format, type, imageBitmaps[type]);
+}
+
+function texImageTypedArray(internalFormat, format, type) {
+  debug("texFunc: texImageTypedArray");
+  gl.texImage2D(gl.TEXTURE_2D, 0, internalFormat, 1, 1, 0, format, type, typedArrays[type]);
+}
 
-  switch(type) {
+function testNorm16Texture(texImageFunc, internalFormat, format, type) {
+  let expectedValue;
+  switch (type) {
     case gl.SHORT:
-      pixelValue = 0x7fff;
-      expectedValue = 0xff;
-      imageData = new Int16Array(4).fill(pixelValue);
+      expectedValue = 0xffff;
       break;
     case gl.UNSIGNED_SHORT:
-      pixelValue = 0x6a35;
-      expectedValue = pixelValue >> 8;
-      imageData = new Uint16Array(4).fill(pixelValue);
+      expectedValue = 0x6a35;
       break;
     default:
       wtu.error("Unreachable: Unknown texture type.");
       break;
   }
 
+  // Render to fbo texture attachment test
+  gl.bindTexture(gl.TEXTURE_2D, textures[1]);
+  gl.texImage2D(gl.TEXTURE_2D, 0, ext.RGBA16_EXT, 1, 1, 0, gl.RGBA, gl.UNSIGNED_SHORT, null);
+
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "rtt bindings succeed");
+
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fbos[0]);
+  gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, textures[1], 0);
+
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "fbo bindings succeed");
+
   // Texture sampled from
   gl.activeTexture(gl.TEXTURE0);
   gl.bindTexture(gl.TEXTURE_2D, textures[0]);
-  gl.texImage2D(gl.TEXTURE_2D, 0, internalFormat, 1, 1, 0, format, type, imageData);
+  texImageFunc(internalFormat, format, type);
 
   wtu.glErrorShouldBe(gl, gl.NO_ERROR, "texture bindings succeed");
 
   gl.drawArrays(gl.TRIANGLES, 0, 6);
 
-  // Read back as gl.UNSIGNED_BYTE
-  wtu.checkCanvasRect(gl, 0, 0, 1, 1, generateFormatColor(format, expectedValue));
+  let readbackBuf = new Uint16Array(4);
+  wtu.checkCanvasRect(gl, 0, 0, 1, 1, generateFormatColor(format, expectedValue, 0xffff), undefined, 0, readbackBuf, gl.UNSIGNED_SHORT);
 }
 
-function testNorm16Render(interalFormat, format, type, tolerance) {
+function testSNorm16TextureImageDisallowed(texImageFunc, internalFormat, format, type) {
+  gl.bindTexture(gl.TEXTURE_2D, textures[0]);
+  texImageFunc(internalFormat, format, type);
+  wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "texture bindings succeed");
```

Comment:
The error message should be "uploads from DOM elements to 16-bit SNORM textures are disallowed" or similar.